chore: Add OpenAPI Support to Statistics API

ahmed-n-abdeltwab opened the pull request:
**Description:**  
This PR integrates OpenAPI support into the `Rocket.Chat API`, migrate of `Rocket.Chat API` endpoints to the new OpenAPI pattern. The update includes improved API documentation, enhanced type safety, and response validation using AJV.

**Key Changes:**  
- **Implemented the new pattern** and added AJV-based JSON schema validation for API.  
- **Uses the ExtractRoutesFromAPI utility** from the TypeScript definitions to dynamically derive the routes from the endpoint specifications.
- **Enabled Swagger UI** integration for this API.
- **This does not introduce any breaking changes** to the endpoint logic.

**Issue Reference:**  
Relates to #34983, part of the ongoing OpenAPI integration effort.

**Testing:**
- Verified that the API response schemas are correctly documented in Swagger UI.  
- All tests passed without any breaking changes

```shell
$ yarn testapi -f '[Statistics]'


  [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (248ms)
      ✔ should return an object with the statistics (224ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (338ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (273ms)
      ✔ should return an array with the statistics (394ms)
      ✔ should return an array with the statistics even requested with count and offset params (202ms)


  6 passing (2s)
```

**Endpoints:**
- [x] statistics:
   - [x] [Get Last Statistics](https://developer.rocket.chat/apidocs/get-last-statistics)

     ![Get Last Statistics](https://github.com/user-attachments/assets/4affb77e-7042-4dbd-b1f0-858ed2db8a20 "Screenshot: Get Last Statistics")
<!--
   - [x] [Get Statistics List](https://developer.rocket.chat/apidocs/get-statistics-list)

     ![Get Statistics List](https://github.com/user-attachments/assets/60d65edb-49fb-496f-b5b8-f0f53fc78750 "Screenshot: Get Statistics List")
-->
   - [x] Statistics Telemetry

     ![Statistics Telemetry](https://github.com/user-attachments/assets/ebadb059-374c-4c86-b360-06012ff81b3b "Screenshot: Statistics Telemetry")


Looking forward to your feedback! 🚀


---

### Description

This pull request introduces OpenAPI support to the statistics API in the `ahmed-n-abdeltwab/Rocket.Chat` repository. The changes are made in the `apps/meteor/app/api/server/v1/stats.ts` file. The update involves refactoring the statistics API routes to incorporate typed validation using ajv schema definitions. This enhancement aims to improve type safety and request validation while preserving the existing functionality of the API. The changes are proposed from the `feat/OpenAPI` branch to be merged into the `develop` branch.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=35692) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=35692) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ahmed-n-abdeltwab<br/>:x: Ahmed N. Abdeltwab<br/><hr/>**Ahmed N. Abdeltwab** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat?pullRequest=35692) it.</sub>

ahmed-n-abdeltwab commented:
@kody start-review

ggazzo commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,54 +1,112 @@
+			response: {
+				200: ajv.compile({
+					type: 'object',
+					properties: {
+						success: { type: 'boolean' },
 					},
```
that's right always use `additionalProperties: false`

ahmed-n-abdeltwab commented:
I usually run `yarn lint -- --fix` before committing, but this time, after running it, I noticed that it applied fixes across the entire project. As a result, when I added my file to Git and committed, it unintentionally included changes from the entire codebase.  

To resolve this, I reverted the unintended changes and manually fixed only my file using the following command:  

```sh
npx eslint --parser @typescript-eslint/parser --parser-options "project: ['./tsconfig.base.json']" apps/meteor/app/api/server/v1/stats.ts --fix
```

This ensured that only the intended file was fixed and committed correctly.  
Should I report this as an issue, or should we wait to see if it happens again?

ahmed-n-abdeltwab commented:
@kody start-review

ahmed-n-abdeltwab commented:
@kody start-review

ahmed-n-abdeltwab commented:
@kody start-review

ahmed-n-abdeltwab commented:
@kody start-review

ahmed-n-abdeltwab commented:
Hi @ggazzo,

I've refactored the statistics endpoints to improve:

1. Code organization (separated handlers from route configs)
2. Schema validation completeness (additionalProperties: false, added descriptions)
3. Documentation readiness (OpenAPI compatibility)

Could you please review these changes and share your feedback.

ggazzo reviewed (changes requested):
there is no reason to "over-optimize" the code.

keep the functions and schemas directly in the endpoint declarations.

this way TypeScript can infer things like `this`, and in this case, since we have several schemas and actions, it is better to understand which one belongs what

ahmed-n-abdeltwab commented:
> there is no reason to "over-optimize" the code.
> 
> keep the functions and schemas directly in the endpoint declarations.
> 
> this way TypeScript can infer things like `this`, and in this case, since we have several schemas and actions, it is better to understand which one belongs what

I see your point! My thought was to improve maintainability by keeping things modular, but I can inline them if that's the preferred structure. Let me know if you want any adjustments!

ahmed-n-abdeltwab commented:
Hey @ggazzo, I've made the recent changes you suggested by keeping functions and schemas inline within the endpoint declarations. Let me know if you have any feedback!

ggazzo commented:
@ahmed-n-abdeltwab  how is your editor configured? we dont need to run the ci to realize we have typescript errors

ahmed-n-abdeltwab commented:
> @ahmed-n-abdeltwab  how is your editor configured? we dont need to run the ci to realize we have typescript errors

Hey @ggazzo, I'm using Vim as my editor. With default configuration

ahmed-n-abdeltwab commented:
> @ahmed-n-abdeltwab how is your editor configured? we dont need to run the ci to realize we have typescript errors

Is there something I need to change in my setup to see the error earlier?

ahmed-n-abdeltwab commented:
@kody start-review

ahmed-n-abdeltwab commented:
> > [/statistics]
> > should return an error when the user does not have the necessary permission:
> > Error: expected 400 "Bad Request", got 502 "Bad Gateway"
> > at /home/runner/work/Rocket.Chat/Rocket.Chat/apps/meteor/tests/end-to-end/api/statistics.ts:19:7
> 
> I think this error happens because the 400 response validation has not been created yet

Hi @ggazzo, you asked for changes! Should I delete my last commit?

ahmed-n-abdeltwab commented:
@ggazzo just pinging you in case you haven't saw my comment

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,197 @@
+			});
+			return API.v1.success({ statistics, success: true });
+		  } catch (error) {
+				return API.v1.failure({ error: error.message }, 400);
```
```
Error: app/api/server/v1/stats.ts(61,36): error TS18046: 'error' is of type 'unknown'.
Error: app/api/server/v1/stats.ts(61,53): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
```

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,197 @@
+			});
+			return API.v1.success({ statistics, success: true });
+      } catch (error) {
+				return API.v1.failure({ error: error.message }, 400);
```
```
Error: app/api/server/v1/stats.ts(122,36): error TS18046: 'error' is of type 'unknown'.
Error: app/api/server/v1/stats.ts(122,53): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
```

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,197 @@
+			}
 			return API.v1.success();
+      } catch (error) {
+				return API.v1.failure({ error: error.message }, 400);
```
```
Error: app/api/server/v1/stats.ts(194,36): error TS18046: 'error' is of type 'unknown'.
Error: app/api/server/v1/stats.ts(194,53): error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
```

cardoso reviewed (changes requested):
Hi @ahmed-n-abdeltwab , we highly recommend you switch to VSCode in order to get instant feedback regarding type errors and eslint errors.

If you still wish to use the command-line:

```
yarn turbo run typecheck
```

```
yarn lint
```

Please ensure both of these commands are successful before sending more commits.

ahmed-n-abdeltwab commented:
> Hi @ahmed-n-abdeltwab , we highly recommend you switch to VSCode in order to get instant feedback regarding type errors and eslint errors.
> 
> If you still wish to use the command-line:
> 
> ```
> yarn turbo run typecheck
> ```
> 
> ```
> yarn lint
> ```
> 
> Please ensure both of these commands are successful before sending more commits.

Got it! I'll check the code and get back to you.

ahmed-n-abdeltwab commented:
The error arises because the status endpoint is set to trigger an error if the user does not have the necessary permissions. Should the `APIClass` internally manage this error by verifying user permissions? I will research to understand what is happening.

cardoso commented:
@ahmed-n-abdeltwab looks like you sent a commit that wasn't tied to your github account:

https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/00b35147320ce94586625ba674a6717e9ca25a86

Please follow the advice in the CLA bot comment:

https://github.com/RocketChat/Rocket.Chat/pull/35692#issuecomment-2774590440

Regarding the permission error you mentioned: whatever the behavior was before, must remain the same. We're only looking to refactor without changing behavior. Maybe @ggazzo coud clarify.

ahmed-n-abdeltwab commented:
Should I revert to this specific commit in Git to attempt a fix, or is it acceptable to leave the current changes as they are for now?

To clarify what I mean by "permission," I'm not referring to user permission. In this context, the user is not me. Instead, I'm talking about whether the user entity is authorized to return the result. In the old code, within the `addRoute` function, the initial logic block checked for permissions and whether authentication was required before proceeding with its main logic, such as querying and performing other tasks. My question is whether the `API` currently performs the same logic to check permissions and handle them accordingly.

ahmed-n-abdeltwab commented:
During my research to identify the problem, I came across the file [`federation.ts`](https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/app/api/server/v1/federation.ts#L8). However, it lacks the logic to check for permissions, which made me question whether what I did is the same as in `federation.ts`, or if I am missing something.

cardoso commented:
@ahmed-n-abdeltwab you should be able to run the tests locally.

This link provides some info: https://developer.rocket.chat/docs/code-style-guide

In the apps/meteor folder, you can run `yarn obj:dev` to start the Rocket.Chat instance.
Check the package.json file in that folder for other scripts such as `test` , `testapi` , `test:e2e`.
I'll leave it to you to figure out what they do and how you can best use them.

> Should I revert to this specific commit in Git to attempt a fix, or is it acceptable to leave the current changes as they are for now?

You can leave as they are for now and focus on fixing the errors.

ahmed-n-abdeltwab commented:
I resolved the issue by adding 400 and 401 response schemas to each endpoint and correcting the query properties according to the documentation from the [website](https://developer.rocket.chat/apidocs/get-last-statistics). Here is the result after running `yarn testapi -f 'statistics'`.
```
[Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (311ms)
      ✔ should return an object with the statistics (227ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (4473ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (195ms)
      ✔ should return an array with the statistics (230ms)
      ✔ should return an array with the statistics even requested with count and offset params (195ms)


  6 passing (6s)
  ```

cardoso commented:
Awesome! Thank you @ahmed-n-abdeltwab ! Just triggered the CI run

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,1020 @@
+						}, // need attentions
+						tag: {
+							anyOf: [{ type: 'null' }, {}],
+						}, //need attentions
```
```suggestion
						}, // need attentions
```

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,1020 @@
+								properties: {
+									id: {
+										anyOf: [{ type: 'null' }, {}],
+									}, //need attentions
```
```suggestion
									}, // need attentions
```

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,1020 @@
+									}, //need attentions
+									alias: {
+										anyOf: [{ type: 'null' }, {}],
+									}, //need attentions
```
```suggestion
									}, // need attentions
```

cardoso commented on apps/meteor/app/api/server/v1/stats.ts:
```diff
@@ -1,62 +1,1020 @@
+									properties: {
+										liveStream: {
+											anyOf: [{ type: 'null' }, {}],
+										}, //need attentions
```
```suggestion
										}, // need attentions
```

cardoso reviewed (changes requested):
Hi @ahmed-n-abdeltwab great job on the summary you just provided. 👏 

Looks like you missed some linting errors:

If you can use VSCode and the eslint and prettier extensions, it should be easier to spot these before committing. If you need a lighter-weight editor maybe try zed.dev which has these built-in.

Otherwise, you can always run `yarn lint`.

Please do this before you push, since the CI won't run the test suite if it fails the lint phase.

I suggested the changes below, but as an exercise I recommend that you make them yourself instead of using github's 'Commit suggestion' button.

ahmed-n-abdeltwab commented:
Thanks a lot! I really appreciated your feedback.

I had already switched my editor to VSCode as you previously recommended, and I installed both the ESLint and Prettier extensions. They definitely helped me catch the linting issues more easily.

I will also run `yarn lint` before pushing and make the suggested changes manually instead of using the 'Commit suggestion' button.

ahmed-n-abdeltwab commented:
### 📊 fix: update comments for clarity

- [x] Updated comments with `FIXME` and `TODO` to clarify current issues
- [x] Ran `Prettier` and `ESLint` in VSCode for formatting and linting

### 🚧 Issues
- `yarn lint` fails due to:
  - ❌ `@rocket.chat/rest-typings` module not found
  - ❌ ESLint cannot resolve the `'hono'` module

Should I investigate this issue or leave it as is?

cardoso commented:
@ahmed-n-abdeltwab great job on the comments 👍

Don't worry about lint issues not related to the files you changed.

ahmed-n-abdeltwab commented:
### 📊 Fix Response Schema for `Get Last Statistics` Endpoint

I used the [`IStats`](https://github.com/RocketChat/Rocket.Chat/blob/feat/OpenAPI/packages/core-typings/dist/IStats.d.ts#L17) interface as a reference to create the schema manually. I thought that by doing so, I could safely use `additionalProperties: false`. But the result was different.

For example, my OS is Debian 12, and the returned `statistics` data includes properties that are not in `IStats`. It seems like the structure changes depending on the OS or environment — like a dynamic interface. Because of this, it's hard to create a static schema. So, for now, I’ve set all `additionalProperties` to `true`.

---

### 🐛 Issues That Caused AJV Validation to Fail

- [`omnichannelSources`](https://github.com/RocketChat/Rocket.Chat/blob/feat/OpenAPI/packages/core-typings/dist/IStats.d.ts#L112): The interface says it's a number or string, but sometimes it returns `undefined`. I’ve hard coded it and left a comment for someone to review. [code](https://github.com/RocketChat/Rocket.Chat/blob/feat/OpenAPI/apps/meteor/app/api/server/v1/stats.ts#L173)

- [`totalLivechatRoomsWithPriority`](https://github.com/RocketChat/Rocket.Chat/blob/feat/OpenAPI/packages/core-typings/dist/IStats.d.ts#L67): The interface marks this field as required, but it’s actually optional in the response. This caused AJV validation to fail, so I commented it out in the `required` array. [code](https://github.com/Rocket.Chat/Rocket.Chat/blob/feat/OpenAPI/apps/meteor/app/api/server/v1/stats.ts#L1028)

- `roomsInsideTeams` and `federationOverviewData`: Same issue as above — they’re marked as required but not always present in the response.

- [lastMessageSentAt](https://github.com/Rocket.Chat/Rocket.Chat/blob/feat/OpenAPI/packages/core-typings/dist/IStats.d.ts#L74): is a JavaScript Date objects and undefined 

- [lockedAt](https://github.com/ahmed-n-abdeltwab/Rocket.Chat/blob/feat/OpenAPI/packages/core-typings/dist/IStats.d.ts#L104): is is a JavaScript Date objects and string.

---

### ✅ Test Results

```bash
$ yarn testapi -f 'statistics'

 [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (323ms)
      ✔ should return an object with the statistics (226ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (754ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (218ms)
      ✔ should return an array with the statistics (194ms)
      ✔ should return an array with the statistics even requested with count and offset params (339ms)

  6 passing (3s)
```

---

### ✅ Note

I also read this blog: [Generating JSON Schema from TypeScript types](https://blog.logrocket.com/generating-json-schema-typescript-types/). It explains a tool called [`ts-json-schema-generator`](https://github.com/vega/ts-json-schema-generator) that can generate JSON Schema directly from TypeScript types.

This tool could help us:
- Avoid writing JSON Schemas manually  
- Keep validation in sync with our TypeScript types  
- Make future updates easier

If that sounds good, let me know what you think! 🙌

ahmed-n-abdeltwab commented:
### 📊 Update Response Schema for `401` Get Last Statistics Endpoint


I added the status and message object to **Response Schema** of `401` for the [`Get Last Statistics`](https://developer.rocket.chat/apidocs/get-last-statistics) endpoint.

no changes from the original docs. [docs link](https://developer.rocket.chat/apidocs/get-last-statistics)


---

### ✅ Test Results

After updating the schema, I ran the tests and all of them passed:

```bash
$ yarn testapi -f 'statistics'

  [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (369ms)
      ✔ should return an object with the statistics (438ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (3060ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (297ms)
      ✔ should return an array with the statistics (350ms)
      ✔ should return an array with the statistics even requested with count and offset params (315ms)


  6 passing (7s)
```

ahmed-n-abdeltwab commented:
### 📊 Enhance Statistics Endpoints

- Integrate Response Schema into the `Get Statistics List` endpoint.
- Add 400 and 401 Response Schema to both `Get Statistics List` and `Statistics Telemetry` endpoints.

#### 🐛 Problems I Found
The interface marks this properties as required, but it’s actually optional in the response. This caused AJV validation to fail, so I commented it out in the required array.
- [totalRoomsWithActiveLivestream](https://github.com/Rocket.Chat/Rocket.Chat/blob/feat/OpenAPI/apps/meteor/app/api/server/v1/stats.ts#L1296)
- [matrixBridgeEnabled](https://github.com/Rocket.Chat/Rocket.Chat/blob/feat/OpenAPI/apps/meteor/app/api/server/v1/stats.ts#L1315)
- dms and channels, groups, teams for [videoConf.properties.settings](https://github.com/Rocket.Chat/Rocket.Chat/blob/feat/OpenAPI/apps/meteor/app/api/server/v1/stats.ts#L874)

> 🔗 I added links to the code to help you check each field.

---

### ✅ Test Results

The tests were executed successfully, and all passed.

```bash
$ yarn testapi -f 'statistics'

  [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (350ms)
      ✔ should return an object with the statistics (254ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (2364ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (224ms)
      ✔ should return an array with the statistics (304ms)
      ✔ should return an array with the statistics even requested with count and offset params (207ms)


  6 passing (5s)
```

---

Looking forward to your feedback! 🚀

ahmed-n-abdeltwab commented:
### 📊 Summary

- [fix: Refactor error response schemas for better reusability](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/b172817bd25c76eb72f67bc52b231c90387e1c71)
- [chore: Comment out unused properties in stats API](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/7aad9d6a095631a3d4a8a34f7b7813d2f3f53df7)
- [refactor: Simplify getLastStatistics call in API response](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/e5389c8a89e94fe70201f4f2e9b004c7e547c4df)
- [refactor: Streamline success response handling in getStatistics API](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/5a03dbd38f7a7fc192481bbc7876005f53046f33)
- [refactor: Replace inline error schemas for Get Statistics List](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/6ba6d8d42d5ed8a01606098a6025b46105ef01db)
- [refactor: Replace inline error schemas for Get Last Statistics](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/66064b048f8df4cc15e27bb5fdc8f73ce51e85fc)
- [refactor: Simplify event handling in telemetry event processing](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/434166fb927f6fb5983c170484423cd885b80deb)
- [refactor: Enhance commonBadRequestErrorSchema with additional properties for improved error handling](https://github.com/RocketChat/Rocket.Chat/pull/35692/commits/3a8cef3bc34e1611ad90b3c1b41d4e728770d3f2)

#### 🐛 Problems I Found
**onHoldEnabled**, **lastChattedAgentPreferred**, **assignNewConversationsToContactManager**, and **visitorAbandonment** have now become unused properties in the stats API 7aad9d6. I had to comment them out to pass the test; I'm still investigating what happened. All of a sudden, they became unused. it wasn't like this yesterday -- it was working just fine.

---

### ✅ Test Results

The tests were executed successfully, and all passed.

```bash
$ yarn testapi -f 'statistics'

 [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (259ms)
      ✔ should return an object with the statistics (213ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (168ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (186ms)
      ✔ should return an array with the statistics (238ms)
      ✔ should return an array with the statistics even requested with count and offset params (281ms)


  6 passing (2s)
```

---

### Github Copilot Reviews
Your changes introduce a robust implementation for handling statistics and telemetry events in the API. Below is a review of your changes, highlighting strengths and areas for improvement:

---

### **Strengths**

1. **Strong Type Safety**:
   - The use of `ConvertedTelemetryMap` ensures that telemetry events are strongly typed, reducing the risk of runtime errors.
   - The `ajv.compile` usage for schema validation is a great way to enforce input validation at runtime.

2. **Comprehensive Schema Definitions**:
   - The `StatisticsSchema` and other schemas are detailed and cover a wide range of properties, ensuring that the API responses are well-structured and predictable.

3. **Dynamic Telemetry Handling**:
   - The `statistics.telemetry` endpoint dynamically handles multiple telemetry event types using the `oneOf` schema. This is a flexible and scalable approach.

4. **Pagination Support**:
   - The `statistics.list` endpoint includes pagination, sorting, and filtering, which are essential for handling large datasets efficiently.

5. **Error Handling**:
   - The inclusion of `commonBadRequestErrorSchema` and `commonUnauthorizedErrorSchema` ensures consistent error responses across endpoints.

6. **Modular Design**:
   - The separation of schemas, validation, and logic makes the code easier to maintain and extend.

---

### **Areas for Improvement**

1. **Error Handling for Telemetry Events**:
   - The `statistics.telemetry` endpoint does not handle cases where `telemetryEvent.call` fails. Consider adding error handling to ensure that failures are logged or reported.

   ```typescript
   try {
       telemetryEvent.call(eventName, params);
   } catch (error) {
       console.error(`Failed to process telemetry event: ${eventName}`, error);
   }
   ```

2. **Validation for `statistics.telemetry`**:
   - The `params` array in the `statistics.telemetry` endpoint is not explicitly validated for length or presence. Add a check to ensure it is not empty or undefined.

   ```typescript
   if (!events?.params?.length) {
       return API.v1.failure({ error: 'No telemetry events provided.' });
   }
   ```

3. **Dynamic Properties in `StatisticsSchema`**:
   - The `additionalProperties: false` is commented out in several places. If the response is dynamic, consider explicitly documenting which properties are dynamic or enabling `additionalProperties: true` where necessary.

4. **Code Duplication in Required Fields**:
   - The `required` arrays in schemas are repeated in multiple places. Consider extracting these into reusable constants to reduce duplication.

   ```typescript
   const requiredStatisticsFields = [
       '_id',
       'wizard',
       'uniqueId',
       'totalUsers',
       'activeUsers',
       // Add other fields here...
   ];
   ```

5. **Telemetry Event Schema**:
   - The `oneOf` schema for telemetry events is hardcoded. If new events are added, this schema will need to be updated manually. Consider dynamically generating the schema based on the `TelemetryMap` type.

6. **Logging**:
   - Add logging for critical operations, such as when telemetry events are processed or when statistics are fetched. This will help with debugging and monitoring.

7. **Performance Considerations**:
   - The `statistics.list` endpoint fetches all statistics matching the query. If the dataset is large, ensure that database queries are optimized with proper indexing and pagination.

8. **Documentation**:
   - Add JSDoc comments for each endpoint to improve readability and provide context for future developers.

   ```typescript
   /**
    * GET /statistics
    * Fetches the latest statistics for the server.
    * @param {boolean} refresh - Whether to refresh the statistics.
    * @returns {object} The server statistics.
    */
   ```

9. **Testing**:
   - Ensure that unit and integration tests are written for all endpoints, especially for edge cases like invalid input or empty telemetry events.

---

### **Example Refactor for `statistics.telemetry`**

Here’s an improved version of the `statistics.telemetry` endpoint:

```typescript
.post(
    'statistics.telemetry',
    {
        authRequired: true,
        body: ajv.compile<{
            params: ConvertedTelemetryMap[];
        }>({
            type: 'object',
            properties: {
                params: {
                    type: 'array',
                    items: {
                        oneOf: [
                            {
                                type: 'object',
                                properties: {
                                    eventName: { const: 'otrStats' },
                                    rid: { type: 'string' },
                                },
                                required: ['eventName', 'rid'],
                                additionalProperties: false,
                            },
                            {
                                type: 'object',
                                properties: {
                                    eventName: { const: 'slashCommandsStats' },
                                    command: { type: 'string' },
                                },
                                required: ['eventName', 'command'],
                                additionalProperties: false,
                            },
                            {
                                type: 'object',
                                properties: {
                                    eventName: { const: 'updateCounter' },
                                    settingsId: { type: 'string' },
                                },
                                required: ['eventName', 'settingsId'],
                                additionalProperties: false,
                            },
                        ],
                    },
                },
            },
            required: ['params'],
            additionalProperties: false,
        }),
        response: {
            200: ajv.compile({
                type: 'object',
                properties: {
                    success: { type: 'boolean' },
                },
                required: ['success'],
                additionalProperties: false,
            }),
            400: ajv.compile(commonBadRequestErrorSchema),
            401: ajv.compile(commonUnauthorizedErrorSchema),
        },
    },
    async function () {
        const { params } = this.bodyParams;

        if (!params?.length) {
            return API.v1.failure({ error: 'No telemetry events provided.' });
        }

        params.forEach((event) => {
            const { eventName, ...eventParams } = event;
            try {
                void telemetryEvent.call(eventName, eventParams);
            } catch (error) {
                console.error(`Failed to process telemetry event: ${eventName}`, error);
            }
        });

        return API.v1.success();
    },
);
```

---

### **Final Thoughts**

Your changes are well-implemented and follow good practices for API design and TypeScript usage. By addressing the areas for improvement, you can further enhance the maintainability, performance, and robustness of your code. Let me know if you need help with specific refactoring or testing!

ahmed-n-abdeltwab commented:
### 📊 Summary

Fix the CI failure by simplifying the required properties in the statistics schema. This attempt is just temporary. until I understand the behavior of the statistics schema; I think it shows different properties when running on different operating systems. I have not tested it yet.

---

### ✅ Test Results

The tests were executed successfully, and all passed.

```bash
$ yarn testapi -f 'statistics'

   [Statistics]
    [/statistics]
      ✔ should return an error when the user does not have the necessary permission (263ms)
      ✔ should return an object with the statistics (249ms)
      ✔ should update the statistics when is provided the "refresh:true" query parameter (187ms)
    [/statistics.list]
      ✔ should return an error when the user does not have the necessary permission (178ms)
      ✔ should return an array with the statistics (245ms)
      ✔ should return an array with the statistics even requested with count and offset params (185ms)


  6 passing (2s)
```

---

ahmed-n-abdeltwab commented on packages/core-typings/src/IStats.ts:
```diff
@@ -99,13 +100,7 @@ export interface IStats {
-		buildAt?: string | Date;
-		lockedAt?: string | Date;
-	};
+	migration: IControl;
```
I checked the IStatistics interface from the server lib to investigate what's going on, and I found out that some properties are causing problems. 
```typescript
// (from apps/meteor/app/statistics/server/lib/statistics.ts lines 418)
statistics.migration = await getControl(); // Promise<IControl>
```
This is one of them. It was missing `hash: string`. So, I updated it to use the exported IControl instead of checking each time something is added to the IControl.

ahmed-n-abdeltwab commented on apps/meteor/app/statistics/server/lib/statistics.ts:
```diff
@@ -222,9 +222,10 @@ export const statistics = {
 		statistics.routingAlgorithm = settings.get('Livechat_Routing_Method');
-
+		// TODO: the onHoldEnabled is duplicated in L205 & L230
 		// is on-hold active
```
the onHoldEnabled, routingAlgorithm are duplicated in the code

cardoso commented:
@ahmed-n-abdeltwab seems tests are failing, but could be unrelated, I'll rerun them.

ahmed-n-abdeltwab commented:
> @ahmed-n-abdeltwab seems tests are failing, but could be unrelated, I'll rerun them.

Thank you so much. By the way, I looked at it before pinging you, and it passed the statistics test, I saw it.

ahmed-n-abdeltwab commented:
im going to update this pr

ahmed-n-abdeltwab commented on apps/meteor/app/statistics/server/lib/getAppsStatistics.ts:
```diff
@@ -8,22 +8,22 @@ import { Info } from '../../../utils/rocketchat.info';
+	totalFailed: number;
+	totalPrivateApps: number;
+	totalPrivateAppsEnabled: number;
 };
```
Numbers can behave like booleans—0 is treated as false and values greater than 0 as true. This causes confusion with the oneOf schemas generated by Typia

ahmed-n-abdeltwab commented on packages/core-typings/src/IStats.ts:
```diff
@@ -67,14 +67,14 @@ export interface IStats {
 	federatedUsers: number;
 	lastLogin: string;
-	lastMessageSentAt: Date | undefined;
+	lastMessageSentAt?: Date;
```
Since both strings and dates are serialized as strings in JSON, Typia's oneOf schema may misinterpret them, leading to validation issues.

ahmed-n-abdeltwab commented on apps/meteor/app/statistics/server/functions/getLastStatistics.ts:
```diff
@@ -12,5 +17,27 @@ export async function getLastStatistics({ userId, refresh }: { userId: IUser['_i
+		stats.createdAt = stats.createdAt.toString();
+	}
+
+	return stats;
```
The date props coming from the database are causing AJV validation errors because they aren't formatted as strings. As a work around fix, I’ve added a check in this function to ensure the dates passed to the controller are properly formatted. no edit or change was made in the logic

ahmed-n-abdeltwab commented on packages/core-typings/src/IStats.ts:
```diff
@@ -74,7 +74,7 @@ export interface IStats {
 	federatedUsers: number;
 	lastLogin: string;
-	lastMessageSentAt: Date | undefined;
+	lastMessageSentAt: string | undefined;
```
The use of Date objects should be forbidden due to JSON serialization issues. Additionally, AJV is configured to fail when parsing Date types, as it expects string-formatted timestamps. the right change was made to fix this bug

ahmed-n-abdeltwab committed:
chore: Add OpenAPI support for the Rocket.Chat Statistics API endpoints by migrating to a modern chained route definition syntax and utilizing shared AJV schemas for validation to enhance API documentation and ensure type safety through response validation.

ahmed-n-abdeltwab commented:
The CI is failing because of the /api/v1/autotranslate.translateMessage endpoint. I have migrated it to the new pattern and discovered a bug in it. Both the migration and the bug fix are included in this PR: #38978

ahmed-n-abdeltwab committed:
Merge branch 'develop' into feat/OpenAPI

ggazzo committed:
Merge branch 'develop' into feat/OpenAPI

ggazzo commented:
Hey @ahmed-n-abdeltwab, thank you for your contribution! 🙏

This endpoint has already been migrated as part of #39553, which consolidated all OpenAPI migrations to make it easier to test and validate everything as a single unit. That PR has been merged into `develop`, so this work is already covered. Closing this one — thanks again!
